File: app/src/main/java/org/puffinbasic/domain/PuffinBasicSymbolTable.java
package org.puffinbasic.domain;

import static org.puffinbasic.domain.STObjects.PuffinBasicAtomTypeId.COMPOSITE;
import static org.puffinbasic.domain.STObjects.PuffinBasicAtomTypeId.DOUBLE;
import static org.puffinbasic.error.PuffinBasicRuntimeError.ErrorCode.BAD_FIELD;
import static org.puffinbasic.error.PuffinBasicRuntimeError.ErrorCode.ILLEGAL_FUNCTION_PARAM;
import static org.puffinbasic.error.PuffinBasicRuntimeError.ErrorCode.MISSING_STRUCT;

import android.os.Build;

import androidx.annotation.RequiresApi;

import org.puffinbasic.domain.STObjects.ArrayReferenceValue;
import org.puffinbasic.domain.STObjects.PuffinBasicAtomTypeId;
import org.puffinbasic.domain.STObjects.PuffinBasicType;
import org.puffinbasic.domain.STObjects.STEntry;
import org.puffinbasic.domain.STObjects.STLValue;
import org.puffinbasic.domain.STObjects.STRef;
import org.puffinbasic.domain.STObjects.STTmp;
import org.puffinbasic.domain.STObjects.STVariable;
import org.puffinbasic.domain.STObjects.StructType;
import org.puffinbasic.domain.Scope.GlobalScope;
import org.puffinbasic.domain.Variable.VariableName;
import org.puffinbasic.error.PuffinBasicInternalError;
import org.puffinbasic.error.PuffinBasicRuntimeError;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;

import it.unimi.dsi.fastutil.chars.Char2ObjectMap;
import it.unimi.dsi.fastutil.chars.Char2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.objects.Object2IntMap;
import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;

@RequiresApi(api = Build.VERSION_CODES.O)
public class PuffinBasicSymbolTable {

    public static final int NULL_ID = -1;
    private final Char2ObjectMap<PuffinBasicAtomTypeId> defaultDataTypes;
    private final Object2ObjectMap<String, StructType> userDefinedTypes;
    private final Object2IntMap<String> labelNameToId;
    private final AtomicInteger idmaker;
    private Scope currentScope;
    private int lastId;
    private int lastLastId;
    private STEntry lastEntry;
    private STEntry lastLastEntry;
    public PuffinBasicSymbolTable() {
        this.defaultDataTypes = new Char2ObjectOpenHashMap<>();
        this.userDefinedTypes = new Object2ObjectOpenHashMap<>();
        this.labelNameToId = new Object2IntOpenHashMap<>();
        this.idmaker = new AtomicInteger();
        this.currentScope = new GlobalScope();
        this.lastId = this.lastLastId = -1;
    }

    private int generateNextId() {
        return idmaker.incrementAndGet();
    }

    public Scope getCurrentScope() {
        return currentScope;
    }

    private Optional<Scope> findScope(Predicate<Scope> predicate) {
        var scope = getCurrentScope();
        while (scope != null) {
            if (predicate.test(scope)) {
                return Optional.of(scope);
            } else {
                scope = scope.getSearchScope();
            }
        }
        return Optional.empty();
    }

    private STEntry getEntry(int id) {
        var scope = getCurrentScope();
        var entry = scope.getNullableEntry(id);
        if (entry != null) {
            return entry;
        } else {
            scope = scope.getParent();
            while (scope != null) {
                entry = scope.getNullableEntry(id);
                if (entry != null) {
                    return entry;
                }
                scope = scope.getParent();
            }
        }
        throw new PuffinBasicInternalError("Failed to find entry for id: " + id);
    }

    public STEntry get(int id) {
        // Cache for better performance
        if (id == lastId) {
            return lastEntry;
        }
        if (id == lastLastId) {
            return lastLastEntry;
        }
        lastLastId = lastId;
        lastLastEntry = lastEntry;
        lastId = id;
        lastEntry = getEntry(id);
        return lastEntry;
    }

    public int getCompositeVariableIdForVariable(VariableName variableName) {
        var scope = findScope(s -> s.containsVariable(variableName)).orElse(getCurrentScope());
        int id = scope.getIdForVariable(variableName);
        if (id == -1) {
            throw new PuffinBasicInternalError("Failed to find variable: " + variableName);
        }
        return id;
    }


    public STEntry getVariable(int id) {
        var entry = get(id);
        if (!entry.isLValue()) {
            throw new PuffinBasicRuntimeError(
                    ILLEGAL_FUNCTION_PARAM,
                    "Entry for id: " + id + " is not a variable"
            );
        }
        return entry;
    }

    public int addVariableOrUDF(
            VariableName variableName,
            Function<VariableName, Variable> variableCreator,
            VariableConsumer consumer) {
        var scope = findScope(s -> s.containsVariable(variableName)).orElse(getCurrentScope());
        int id = scope.getIdForVariable(variableName);
        final STVariable entry;
        if (id == -1) {
            id = generateNextId();
            scope.putVariable(variableName, id);
            var variable = variableCreator.apply(variableName);
            entry = variableName.getDataType().createVariableEntry(variable);
            scope.putEntry(id, entry);
        } else {
            entry = (STVariable) get(id);
        }
        consumer.consume(id, entry, entry.getVariable());
        return id;
    }

    public int addCompositeVariable(
            VariableName variableName,
            STVariable variable) {
        var scope = findScope(s -> s.containsVariable(variableName)).orElse(getCurrentScope());
        int id = generateNextId();
        scope.putVariable(variableName, id);
        scope.putEntry(id, variable);
        return id;
    }

    public int addLabel(String label) {
        var id = labelNameToId.getOrDefault(label, -1);
        if (id == -1) {
            id = addLabel();
            labelNameToId.put(label, id);
        }
        return id;
    }

    public int addLabel() {
        var scope = getCurrentScope();
        var id = generateNextId();
        var entry = new STObjects.STLabel();
        scope.putEntry(id, entry);
        return id;
    }

    public int addGotoTarget() {
        var scope = getCurrentScope();
        int id = generateNextId();
        var entry = PuffinBasicAtomTypeId.INT32.createTmpEntry();
        scope.putEntry(id, entry);
        return id;
    }

    public int addArrayReference(STLValue lvalue) {
        var ref = new ArrayReferenceValue(lvalue);
        int id = generateNextId();
        var entry = new STLValue(ref, lvalue.getType());
        getCurrentScope().putEntry(id, entry);
        return id;
    }

    public int addTmp(PuffinBasicType type, Consumer<STEntry> consumer) {
        var scope = getCurrentScope();
        int id = generateNextId();
        var entry = type.canBeLValue() ? new STLValue(null, type) : new STTmp(null, type);
        entry.createAndSetInstance(this);
        scope.putEntry(id, entry);
        consumer.accept(entry);
        return id;
    }

    public int addTmp(PuffinBasicAtomTypeId dataType, Consumer<STEntry> consumer) {
        var scope = getCurrentScope();
        int id = generateNextId();
        var entry = dataType.createTmpEntry();
        scope.putEntry(id, entry);
        consumer.accept(entry);
        return id;
    }

    public int addRef(PuffinBasicType type) {
        var scope = getCurrentScope();
        int id = generateNextId();
        var entry = new STRef(type);
        scope.putEntry(id, entry);
        return id;
    }

    public int addTmpCompatibleWith(int srcId) {
        var scope = getCurrentScope();
        var dataType = scope.getEntry(srcId).getType().getAtomTypeId();
        int id = generateNextId();
        scope.putEntry(id, dataType.createTmpEntry());
        return id;
    }

    public PuffinBasicAtomTypeId getDataTypeFor(String varname, String suffix) {
        var scope = getCurrentScope();
        if (scope.containsVariable(new VariableName(varname, null, COMPOSITE))) {
            return COMPOSITE;
        }
        if (varname.length() == 0) {
            throw new PuffinBasicInternalError("Empty variable name: " + varname);
        }
        if (suffix == null) {
            var firstChar = varname.charAt(0);
            return defaultDataTypes.getOrDefault(firstChar, DOUBLE);
        } else {
            return PuffinBasicAtomTypeId.lookup(suffix);
        }
    }

    public void setDefaultDataType(char c, PuffinBasicAtomTypeId dataType) {
        defaultDataTypes.put(c, dataType);
    }

    public void addStructType(String name, StructType type) {
        userDefinedTypes.put(name, type);
    }

    public void checkUnused(String name) {
        if (userDefinedTypes.containsKey(name)) {
            throw new PuffinBasicRuntimeError(
                    BAD_FIELD,
                    "Name: " + name + " is already used!"
            );
        }
    }

    public StructType getStructType(String name) {
        var type = userDefinedTypes.get(name);
        if (type == null) {
            throw new PuffinBasicRuntimeError(
                    MISSING_STRUCT,
                    "Missing struct: " + name
            );
        }
        return type;
    }

    public void pushDeclarationScope(int funcId, boolean localScope) {
        currentScope = getCurrentScope().createChild(funcId, localScope);
    }

    public void pushRuntimeScope(int funcId, int callerInstrId) {
        var funcDeclScope = getCurrentScope().getChild(funcId);
        if (funcDeclScope == null) {
            throw new PuffinBasicInternalError("Failed to find scope for id: " + funcId);
        }
        currentScope = funcDeclScope.createRuntimeScope(callerInstrId);
    }

    public void popScope() {
        var parent = getCurrentScope().getParent();
        if (parent == null) {
            throw new PuffinBasicInternalError("Scope underflow!");
        }
        currentScope = parent;
    }

    public interface VariableConsumer {
        void consume(int id, STVariable entry, Variable variable);
    }
}

File: app/src/main/java/org/puffinbasic/domain/PuffinBasicSymbolTable.kt
package org.puffinbasic.domain

import org.puffinbasic.domain.STObjects.ArrayReferenceValue
import org.puffinbasic.domain.STObjects.PuffinBasicAtomTypeId
import org.puffinbasic.domain.STObjects.PuffinBasicType
import org.puffinbasic.domain.STObjects.STEntry
import org.puffinbasic.domain.STObjects.STLValue
import org.puffinbasic.domain.STObjects.STLabel
import org.puffinbasic.domain.STObjects.STRef
import org.puffinbasic.domain.STObjects.STTmp
import org.puffinbasic.domain.STObjects.STVariable
import org.puffinbasic.domain.STObjects.StructType
import org.puffinbasic.domain.Variable.VariableName
import org.puffinbasic.error.PuffinBasicInternalError
import org.puffinbasic.error.PuffinBasicRuntimeError
import java.util.Optional
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate

class PuffinBasicSymbolTable {
    private val defaultDataTypes: MutableMap<Char, PuffinBasicAtomTypeId> = mutableMapOf()
    private val userDefinedTypes: MutableMap<String, StructType> = mutableMapOf()
    private val labelNameToId: MutableMap<String, Int> = mutableMapOf()
    private val idMaker: AtomicInteger = AtomicInteger()
    var currentScope: Scope?
        private set
    private var lastId: Int
    private var penultimateId: Int
    private var lastEntry: STEntry? = null
    private var penultimateEntry: STEntry? = null

    init {
        currentScope = GlobalScope()
        penultimateId = -1
        lastId = penultimateId
    }

    private fun generateNextId() = idMaker.incrementAndGet()

    private fun findScope(predicate: Predicate<Scope>): Optional<Scope> {
        var scope = currentScope
        while (scope != null) {
            scope = if (predicate.test(scope)) {
                return Optional.of(scope)
            } else {
                scope.searchScope
            }
        }
        return Optional.empty()
    }

    private fun getEntry(id: Int): STEntry {
        var scope = currentScope
        var entry = scope!!.getNullableEntry(id)
        if (entry != null) {
            return entry
        } else {
            scope = scope.parent
            while (scope != null) {
                entry = scope.getNullableEntry(id)
                if (entry != null) {
                    return entry
                }
                scope = scope.parent
            }
        }
        throw PuffinBasicInternalError("Failed to find entry for id: $id")
    }

    operator fun get(id: Int): STEntry? {
        // Cache for better performance
        if (id == lastId) {
            return lastEntry
        }
        if (id == penultimateId) {
            return penultimateEntry
        }
        penultimateId = lastId
        penultimateEntry = lastEntry
        lastId = id
        lastEntry = getEntry(id)
        return lastEntry
    }

    fun getCompositeVariableIdForVariable(variableName: VariableName): Int {
        val scope = findScope { s: Scope -> s.containsVariable(variableName) }.orElse(
            currentScope
        )
        val id = scope!!.getIdForVariable(variableName)
        if (id == -1) {
            throw PuffinBasicInternalError("Failed to find variable: $variableName")
        }
        return id
    }

    fun getVariable(id: Int): STEntry {
        val entry = get(id)
        if (!entry!!.isLValue) {
            throw PuffinBasicRuntimeError(
                PuffinBasicRuntimeError.ErrorCode.ILLEGAL_FUNCTION_PARAM,
                "Entry for id: $id is not a variable"
            )
        }
        return entry
    }

    fun addVariableOrUDF(
        variableName: VariableName,
        variableCreator: Function<VariableName?, Variable>,
        consumer: VariableConsumer
    ): Int {
        val scope = findScope { s: Scope -> s.containsVariable(variableName) }.orElse(
            currentScope
        )
        var id = scope!!.getIdForVariable(variableName)
        val entry: STVariable?
        if (id == -1) {
            id = generateNextId()
            scope.putVariable(variableName, id)
            val variable = variableCreator.apply(variableName)
            entry = variableName.dataType.createVariableEntry(variable)
            scope.putEntry(id, entry)
        } else {
            entry = get(id) as STVariable
        }
        consumer.consume(id, entry, entry.variable)
        return id
    }

    fun addCompositeVariable(
        variableName: VariableName,
        variable: STVariable
    ): Int {
        val scope = findScope { s: Scope -> s.containsVariable(variableName) }.orElse(
            currentScope
        )
        val id = generateNextId()
        scope!!.putVariable(variableName, id)
        scope.putEntry(id, variable)
        return id
    }

    fun addLabel(label: String): Int {
        var id: Int = labelNameToId.getOrDefault(label, -1)
        if (id == -1) {
            id = addLabel()
            labelNameToId[label] = id
        }
        return id
    }

    fun addLabel(): Int {
        val scope = currentScope
        val id = generateNextId()
        val entry = STLabel()
        scope!!.putEntry(id, entry)
        return id
    }

    fun addGotoTarget(): Int {
        val scope = currentScope
        val id = generateNextId()
        val entry = PuffinBasicAtomTypeId.INT32.createTmpEntry()
        scope!!.putEntry(id, entry)
        return id
    }

    fun addArrayReference(lvalue: STLValue): Int {
        val ref = ArrayReferenceValue(lvalue)
        val id = generateNextId()
        val entry = STLValue(ref, lvalue.type)
        currentScope!!.putEntry(id, entry)
        return id
    }

    fun addTmp(type: PuffinBasicType, consumer: Consumer<STEntry>): Int {
        val scope = currentScope
        val id = generateNextId()
        val entry = if (type.canBeLValue()) STLValue(null, type) else STTmp(null, type)
        entry.createAndSetInstance(this)
        scope!!.putEntry(id, entry)
        consumer.accept(entry)
        return id
    }

    fun addTmp(dataType: PuffinBasicAtomTypeId, consumer: Consumer<STEntry?>): Int {
        val scope = currentScope
        val id = generateNextId()
        val entry = dataType.createTmpEntry()
        scope!!.putEntry(id, entry)
        consumer.accept(entry)
        return id
    }

    fun addRef(type: PuffinBasicType?): Int {
        val scope = currentScope
        val id = generateNextId()
        val entry = STRef(type)
        scope!!.putEntry(id, entry)
        return id
    }

    fun addTmpCompatibleWith(srcId: Int): Int {
        val scope = currentScope
        val dataType = scope!!.getEntry(srcId)!!.type!!.atomTypeId
        val id = generateNextId()
        scope.putEntry(id, dataType.createTmpEntry())
        return id
    }

    fun getDataTypeFor(varname: String, suffix: String?): PuffinBasicAtomTypeId {
        val scope = currentScope
        if (scope!!.containsVariable(
                VariableName(
                    varname,
                    null,
                    PuffinBasicAtomTypeId.COMPOSITE
                )
            )
        ) {
            return PuffinBasicAtomTypeId.COMPOSITE
        }
        if (varname.isEmpty()) {
            throw PuffinBasicInternalError("Empty variable name: $varname")
        }
        return if (suffix == null) {
            val firstChar = varname[0]
            defaultDataTypes.getOrDefault(firstChar, PuffinBasicAtomTypeId.DOUBLE)
        } else {
            PuffinBasicAtomTypeId.lookup(suffix)
        }
    }

    fun setDefaultDataType(c: Char, dataType: PuffinBasicAtomTypeId) {
        defaultDataTypes[c] = dataType
    }

    fun addStructType(name: String, type: StructType) {
        userDefinedTypes[name] = type
    }

    fun checkUnused(name: String) {
        if (userDefinedTypes.containsKey(name)) {
            throw PuffinBasicRuntimeError(
                PuffinBasicRuntimeError.ErrorCode.BAD_FIELD,
                "Name: $name is already used!"
            )
        }
    }

    fun getStructType(name: String): StructType {
        return userDefinedTypes[name]
            ?: throw PuffinBasicRuntimeError(
                PuffinBasicRuntimeError.ErrorCode.MISSING_STRUCT,
                "Missing struct: $name"
            )
    }

    fun pushDeclarationScope(funcId: Int, localScope: Boolean) {
        currentScope = currentScope!!.createChild(funcId, localScope)
    }

    fun pushRuntimeScope(funcId: Int, callerInstrId: Int) {
        val funcDeclScope = currentScope!!.getChild(funcId)
        currentScope = funcDeclScope.createRuntimeScope(callerInstrId)
    }

    fun popScope() {
        val parent = currentScope!!.parent ?: throw PuffinBasicInternalError("Scope underflow!")
        currentScope = parent
    }

    interface VariableConsumer {
        fun consume(id: Int, entry: STVariable, variable: Variable)
    }

    companion object {
        const val NULL_ID = -1
    }
}

fun variableConsumer(consumer: (id: Int, entry: STVariable, variable: Variable) -> Unit): PuffinBasicSymbolTable.VariableConsumer {
    return object : PuffinBasicSymbolTable.VariableConsumer {
        override fun consume(id: Int, entry: STVariable, variable: Variable) {
            consumer(id, entry, variable)
        }
    }
}